Feature unavailable unified error support

RonEld opened the pull request:
## Description
This PR adds support for `MBEDTLS_ERR_PLATFORM_FEATURE_UNSUPPORTED` introduced in #2054 
Tests that receive this error, skip to the next step, as the underlying HW accelerator does not support that specific feature.
Exception are the entropy, when the SHA function returns such an error, as the specific entropy source is dependent on the SHA function, and not configurable, and the ctr_drbg, when the specific AES functionality is not supported, as it is dependent on it to work. Skipping the tests may result in security issue, if developer assumes the test passes.
Second PR that supersedes #1716
## Status
**READY**

## Requires Backporting
NO  


## Migrations
~Error codes have been removed, so we may consider this as an API change~

## Additional comments

ctr_drbg requires `MBEDTLS_CTR_DRBG_USE_128_BIT_KEY` to be defined, in case AES 256 is not supported by the HW, ~once PR #1902 will be merged.~

~Note: Once #2054 will be merged, I will rebase to have the support for the on target testing to run, but this PR can start the review~

## Todos
- [x] Tests
- [x] Documentation
- [ ] Changelog updated
- [ ] Backported

RonEld commented:
Note: documentation updated in https://github.com/ARMmbed/mbed-tls-docs/pull/112

RonEld commented:
updated changeLog, although I am not sure this change in tests and the benchmark application requires a ChangeLog entry.

RonEld commented:
#2054 has been merged, so no more waiting for preceding PR

RonEld commented:
After rebase I have tested on PC, to check nothing got broken.

I am currently building on target with unsupported features, and fix if new issues arise from the rebase ( such as new tests introduced that require the new macro)

gilles-peskine-arm commented on library/ecdsa.c:
```diff
@@ -752,8 +752,12 @@ int mbedtls_ecdsa_read_signature_restartable( mbedtls_ecdsa_context *ctx,
-            mbedtls_ecp_gen_keypair( &ctx->grp, &ctx->d, &ctx->Q, f_rng, p_rng ) );
+    int ret = 0;
+    ret = mbedtls_ecp_group_load( &ctx->grp, gid );
+    if( ret == 0 )
```
Minor: I think it would be more readable to use the common idiom `if( ret != 0 ) return( ret );`

gilles-peskine-arm commented on library/ecdsa.c:
```diff
@@ -752,8 +752,12 @@ int mbedtls_ecdsa_read_signature_restartable( mbedtls_ecdsa_context *ctx,
 {
-    return( mbedtls_ecp_group_load( &ctx->grp, gid ) ||
-            mbedtls_ecp_gen_keypair( &ctx->grp, &ctx->d, &ctx->Q, f_rng, p_rng ) );
+    int ret = 0;
```
Oooh, the old code had a bug and the fix should be backported. `f() || g()` returned 1 on failure. This broke application code that looks for errors with `if (mbedtls_ecdsa_genkey(…) < 0)`, which should be correct because the library normally returns negative values on error. This should also be mentioned in the changelog as a bugfix.

RonEld replied:
Backported to 2.7 in https://github.com/ARMmbed/mbedtls/pull/2296

gilles-peskine-arm commented on library/pkparse.c:
```diff
@@ -529,7 +530,11 @@ static int pk_get_rsapubkey( unsigned char **p,
                                         NULL, 0, NULL, 0 ) ) != 0 )
-        return( MBEDTLS_ERR_PK_INVALID_PUBKEY );
+    {
+        if( ret != MBEDTLS_ERR_PLATFORM_FEATURE_UNSUPPORTED )
```
The documentation in `pk.h` needs to be updated to add `MBEDTLS_ERR_PLATFORM_FEATURE_UNSUPPORTED` as a possible error where applicable.

RonEld replied:
I disagree that we should mention generic error codes in our function description, however, I'm willing to add the  generic platform errors are possible statement (`MBEDTLS_ERR_PLATFORM_HW_ACCEL_FAILED` is also an option)

gilles-peskine-arm commented on library/pkparse.c:
```diff
@@ -1343,7 +1361,9 @@ int mbedtls_pk_parse_key( mbedtls_pk_context *pk,
      * on failed PEM parsing attempts. */
 
-    return( MBEDTLS_ERR_PK_KEY_INVALID_FORMAT );
+    if( ret & MBEDTLS_ERR_PLATFORM_FEATURE_UNSUPPORTED )
```
Masking an error code doesn't make sense. Did you mean “if the low-level error is feature-unsupported then mask out the high-level error”? That would be
```
    if( ( ret & -0x7f ) == MBEDTLS_ERR_PLATFORM_FEATURE_UNSUPPORTED )
```

gilles-peskine-arm commented on library/pkparse.c:
```diff
@@ -1343,7 +1361,9 @@ int mbedtls_pk_parse_key( mbedtls_pk_context *pk,
 
-    return( MBEDTLS_ERR_PK_KEY_INVALID_FORMAT );
+    if( ret & MBEDTLS_ERR_PLATFORM_FEATURE_UNSUPPORTED )
+        return( MBEDTLS_ERR_PLATFORM_FEATURE_UNSUPPORTED );
```
Why does `MBEDTLS_ERR_PLATFORM_FEATURE_UNSUPPORTED` get this special treatment of masking out the high-level error? And similarly in other places why does it get the special treatment of not masking in a high-level error?

RonEld commented on programs/test/benchmark.c:
```diff
@@ -884,18 +1062,41 @@ int main( int argc, char *argv[] )
             mbedtls_ecdh_init( &ecdh );
+            mbedtls_snprintf( title, sizeof( title ), "ECDHE-%s",
+                                              curve_info->name );
+            if( mbedtls_ecp_group_load( &ecdh.grp, curve_info->grp_id ) != 0 )
```
ok

RonEld commented on programs/test/benchmark.c:
```diff
@@ -136,7 +135,12 @@ do {                                                                    \
-    if( ret != 0 )                                                      \
+    if( ret == MBEDTLS_ERR_PLATFORM_FEATURE_UNSUPPORTED )               \
+    {                                                                   \
+        mbedtls_printf( "Feature Not Supported. Skipping.\n" );         \
```
ok

RonEld commented on tests/suites/helpers.function:
```diff
@@ -64,11 +63,30 @@ typedef struct data_tag
 
+#define TEST_ASSERT_RET( TEST, EXP_RET )                                              \
+    do {                                                                              \
+        int ret = (TEST);                                                             \
```
yes, this is why I renamed all `ret` in the tests to something like `exp_ret`

RonEld commented on tests/suites/helpers.function:
```diff
@@ -64,11 +63,30 @@ typedef struct data_tag
 /*----------------------------------------------------------------------------*/
 /* Macros */
 
+#define TEST_ASSERT_RET( TEST, EXP_RET )                                              \
```
This macro specifically tests that the return code of the functoin is as expected. Thus the name `TEST_ASSERT_RET`.
If the function returns `MBEDTLS_ERR_PLATFORM_FEATURE_UNSUPPORTED`, then we skip the test.
This Macro should probably be combined with [TEST_INVALID_PARAM_RET](https://github.com/ARMmbed/mbedtls/pull/1643/files#diff-5c77e3830486ba0964d963b0b85814ddR132)

RonEld commented on library/x509_crt.c:
```diff
@@ -1320,7 +1325,15 @@ int mbedtls_x509_crt_parse_path( mbedtls_x509_crt *chain, const char *path )
+        /*
+         * Quit parsing on a feature not suppoprted by HW error.
+         */
+        if( w_ret == MBEDTLS_ERR_PLATFORM_FEATURE_UNSUPPORTED )
```
Because we need to know if the error was due to a feature unsupported error or any other error.
This function returns the number of certificates failed to parse. In a case that one of the certificates failed because of unsupported feature, we need a way to identify, and skip the tests. Note this is still a failure.

RonEld commented on library/x509_crt.c:
```diff
@@ -1842,20 +1865,37 @@ static int x509_crt_verifycrl( mbedtls_x509_crt *crt, mbedtls_x509_crt *ca,
-            /* Note: this can't happen except after an internal error */
+            if( ret == MBEDTLS_ERR_PLATFORM_FEATURE_UNSUPPORTED )
+            {
+                flags |= MBEDTLS_X509_BADCRL_ALG_NOT_SUPPORTED;
```
Because we can't verify the certificate, it is still not trusted, but we still want to know that it was due to an algorithm not supported. As mentioned by @mpg in https://github.com/ARMmbed/mbedtls/pull/2124#pullrequestreview-184622147:
>I particularly like that the new flags are set in addition to the existing NOT_TRUSTED flags rather than replacing them. I think the minimizes the risk of misuse and is an excellent choice.

RonEld commented:
@gilles-peskine-arm Thank you for the review.

I addressed your comments, and will make the fixes shortly.

As for your general comments:
>This PR makes some functions return an unadorned MBEDTLS_ERR_PLATFORM_FEATURE_UNSUPPORTED when other low-level errors are combined with a high-level error. 

I did this only when the low level flags are `MBEDTLS_ERR_PLATFORM_FEATURE_UNSUPPORTED`. I don't see a case where there is more than one low level flag.

>I'm mostly ok with the test case skipping feature, but it should use a tristate, not two booleans. Please take a look at #1308 which introduced a test case skipping feature to see if there would be problems reconciling them. Do we want a different message for “skipped because of an unsupported feature in the hardware” and “skipped because of an unmet assumption”? Cc @hanno-arm

Ok, I'll change. Actually, The behavior should be the same for both PRs. This PR already handle rsa feature unsupported as well. I believe in rsa_alt case as well. `MBEDTLS_ERR_PLATFORM_FEATURE_UNSUPPORTED` is returned by the developer implementing the alternative implementation, for what any reason they choose.

>The change in the X509 module needs design approval from @sbutcher-arm or someone delegated by him.

As you can see from https://github.com/ARMmbed/mbedtls/pull/2124#pullrequestreview-184622147 , @mpg already reviewed and approved the changes in the x509 module.

NirSonnenschein commented on include/mbedtls/x509.h:
```diff
@@ -85,27 +85,28 @@
-#define MBEDTLS_X509_BADCRL_BAD_PK           0x040000  /**< The CRL is signed with an unacceptable PK alg (eg RSA vs ECDSA). */
-#define MBEDTLS_X509_BADCRL_BAD_KEY          0x080000  /**< The CRL is signed with an unacceptable key (eg bad curve, RSA too short). */
-
+#define MBEDTLS_X509_BADCERT_EXPIRED               0x01  /**< The certificate validity has expired. */
```
ok

RonEld commented on library/pkparse.c:
```diff
@@ -539,14 +544,20 @@ static int pk_get_rsapubkey( unsigned char **p,
 
-    if( mbedtls_rsa_complete( rsa ) != 0 ||
-        mbedtls_rsa_check_pubkey( rsa ) != 0 )
+    if( ( ret = mbedtls_rsa_complete( rsa ) ) != 0 ||
```
This is a Boolean condition. If the first statement fails, then we return the first `ret`, if the second statement fails then the second `ret` is returned

RonEld replied:
scratch that

RonEld replied:
Actually, don't scratch it. If the first statement is true ( `ret != 0` ) then we don't call the second statement

RonEld commented on library/rsa.c:
```diff
@@ -664,10 +664,11 @@ int mbedtls_rsa_check_privkey( const mbedtls_rsa_context *ctx )
+        ( ret = mbedtls_rsa_check_privkey( prv ) ) != 0 )
     {
-        return( MBEDTLS_ERR_RSA_KEY_CHECK_FAILED );
+        return( ret );
```
if the first statement fails, then the second statement is not even called

RonEld replied:
scratch that..

RonEld replied:
Actually, don't scratch it. If the first statement is true ( `ret != 0` ) then we don't call the second statement

RonEld commented:
@NirSonnenschein I have modified according to your comment, in order to make the code more readable.

Note that the original change in `rsa` actually changed behaviour, since instead of returning `MBEDTLS_ERR_RSA_KEY_CHECK_FAILED ` , the error code from the `mbedtls_rsa_check_xxxkey` was returned. I reverted this change of behavior, returning an addition ` MBEDTLS_ERR_PLATFORM_FEATURE_UNSUPPORTED` where needed

RonEld commented on library/ccm.c:
```diff
@@ -439,7 +438,14 @@ int mbedtls_ccm_self_test( int verbose )
+    if( ret == MBEDTLS_ERR_PLATFORM_FEATURE_UNSUPPORTED )
+    {
+        if( verbose != 0 )
+            mbedtls_printf( "  CCM: skipped\n" );
```
Actually, I think all the others should be "CCM: skipped". How do we know that it was a CCM failure?

RonEld replied:
I take it back...
The other places are after the print:
~~~
 CCM-AES #<iteration>
~~~
so we know it's a CCM failure, and we don't want a message like:
~~~
CCM-AES #1 CCM: skipped
~~~

However, this print is before we know it's a CCM test.
Bottom line, current logs are the desired behavior, IMHO

sbutcher-arm commented:
Sorry, Ron. This looks like it needs rebasing after PR #1643.

@RonEld - You have a timezone advantage, if you can rework the PR, and get approval before around 10am, I can still merge this.

mpg commented:
@RonEld If it helps, I'm available to review the result of the rebase (conflict resolution) based on the previous approvals.

RonEld commented:
@sbutcher-arm It won't make it by 10am. It needs building and testing on target, which takes a few hours

RonEld commented:
@mpg Thanks you for your suggestion, but I really need to test on target, as I believe that #1643 broke this feature, and this means some rework on this feature.

RonEld commented:
@sbutcher-arm  I have rebased to resolve conflicts.

@gilles-peskine-arm The commits addressing your comments start at https://github.com/ARMmbed/mbedtls/pull/2124/commits/a74aa274e318d22a015fdc5f1a48cb85dfecd76f

I have also modified the Invalid params test to the unsupported feature error. cc @k-stachowiak 

cc @moranpeker @NirSonnenschein

sbutcher-arm commented:
@moranpeker @NirSonnenschein @mpg @gilles-peskine-arm - Could you please re-review this PR for Ron?

RonEld commented on tests/suites/test_suite_rsa.function:
```diff
@@ -654,17 +657,12 @@ void rsa_pkcs1_verify_raw( data_t * hash_result,
+                                                         &olen, result_str->x,
+                                                         output, sizeof( output ) ) ),
+                                                         0 );
         if( res == 0 )
```
`res` could also be `MBEDTLS_ERR_PLATFORM_FEATURE_UNSUPPPORTED`, but in that case, we would skip the test. I'll remove this check.

RonEld commented on tests/suites/test_suite_rsa.function:
```diff
@@ -654,17 +657,12 @@ void rsa_pkcs1_verify_raw( data_t * hash_result,
+                                                         0 );
         if( res == 0 )
         {
             ok = olen == hash_result->len && memcmp( output, hash_result->x, olen ) == 0;
```
perhaps, but this deviates from the purpose of this PR, and this PR is already too big.

RonEld replied:
I'll make the change

gilles-peskine-arm commented:
Unfortunately this needs rework since it has several conflicts. And maybe this should be staged into a crypto PR first and later a TLS PR?

RonEld commented:
Yes, due to the long time, it should probably be done now in the crypto PR

RonEld commented:
The crypto tests and code has been staged as a crypto PR in https://github.com/ARMmbed/mbed-crypto/pull/162

RonEld commented:
Rebased to resolve conflicts

mpg commented:
This PR is about improving support for `_ALT`, which is now on its way out (being replaced with the PSA Crypto drvier API). It has gotten a lot of conflicts over the years, so reviving it would be more effort that warranted considering `_ALT` support is no longer a priority.
